feat: add remix guide and starter repo (#60)
teamhanko/docs: quickstarts/fullstack/remix.mdx

---

## Install `@teamhanko/hanko-elements`

Once you've initialized your Remix app, installing hanko-elements provides you with access to the prebuilt components: `hanko-auth` and `hanko-profile`.

<CodeGroup>
```bash npm
npm install @teamhanko/hanko-elements
```

```bash pnpm
pnpm add @teamhanko/hanko-elements
```

```bash yarn
yarn add @teamhanko/hanko-elements
```
</CodeGroup>

## Add the Hanko API URL

Retrieve the API URL from the [Hanko console](https://cloud.hanko.io/) and place it in your `.env` file.

```sh .env
HANKO_API_URL=https://f4****-4802-49ad-8e0b-3d3****ab32.hanko.io
```

<Note>
  If you are self-hosting you need to provide the URL of your running Hanko
  backend.
</Note>

## Importing and exporting Hanko functions in a `.client.ts` file

Since the `@teamhanko/hanko-elements` package is designed to work on the client side, you'll need to import the Hanko functions into a `.client.ts` file. This file will then export them, making them available for use throughout your application.

```tsx utils/hanko.client.ts
import { register, Hanko } from "@teamhanko/hanko-elements";

export { register, Hanko };
```

## Add `<hanko-auth>` component

> The `<hanko-auth>` web component adds a login interface to your app. This code defines a `HankoAuth` component in the `login` route. 

Reviewer: ```suggestion
The `<hanko-auth>` web component adds a login interface to your app. The following code example defines a `HankoAuth` component in the `login` route. 
```

> It uses the register function from `hanko.client.ts` to register Hanko with the provided URL. On successful registration, it sets the hanko state. When hanko is set, it sets up a callback for when the authentication flow is completed, redirecting the user to the dashboard.

Reviewer: > " ... to register Hanko with the provided URL."

I feel that, technically, this is a bit inaccurate. What it does is register a custom element with the custom element registry of the browser.

Author: Modified it to this, please let me know

The `<hanko-auth>` web component adds a login interface to your app. The following code example defines a `HankoAuth` component in the `login` route. 

It uses the register function from `hanko.client.ts` to register `<hanko-auth>` component with the brower's [CustomElementRegistry](https://developer.mozilla.org/de/docs/Web/API/CustomElementRegistry). When the authentication flow is completed, a callback is set up to redirect the user to the dashboard.

> It uses the register function from `hanko.client.ts` to register Hanko with the provided URL. On successful registration, it sets the hanko state. When hanko is set, it sets up a callback for when the authentication flow is completed, redirecting the user to the dashboard.

Reviewer: >  it sets the hanko state

Again, I feel this is too vague. As a user, I wouldn't even know what "the hanko state is". Maybe give a short explain what `Hanko` actually is in this case?

            .catch((error: Error) => {
                console.error(error.message);
            })
            .then((result) => {
                if (result) {
                    setHanko(result.hanko);
                }
            });
    }, [hankoUrl]);

    // here we define the logout function
    const logout = async () => {
        try {
            await hanko?.user.logout();
            navigate("/login");
            return;
        } catch (error) {
            console.error("Error during logout:", error);
        }
    };

    return (
        <div className="">
            <Suspense fallback={"Loading..."}>
                <hanko-profile />
            </Suspense>
            <div>
                <button onClick={logout}>Logout</button>
            </div>
        </div>
    )
}

const Dashboard = () => {
    return (
        <div>
            <HankoProfile />
        </div>
    )
}

export default Dashboard
```

## Customize component styles

You can customize the appearance of `hanko-auth` and `hanko-profile` components using CSS variables and parts. Refer to our [customization guide](/guides/hanko-elements/customize-appearance).

## Securing routes

> To add JWT verification with Hanko in your Remix application, we're using [jose library](https://www.npmjs.com/package/jose). However, you're free to choose any other suitable library. 

Reviewer: ```suggestion
To add JWT verification with Hanko in your Remix application, we're using the [jose](https://www.npmjs.com/package/jose) library. However, you're free to choose any other suitable library. 
```

Below we define a `validateJwtAndFetchUserId` function which validates the JWT and fetches the user ID. This will also be used to protect routes and ensure only authenticated users can access certain parts of the application.

```jsx app/services/auth.server.ts
import { parse } from "cookie";
import { jwtVerify, createRemoteJWKSet } from "jose";

const hankoApiUrl = process.env.HANKO_API_URL;

export async function validateJwtAndFetchUserId(request: Request) {
  const cookies = parse(request.headers.get("Cookie") || "");
  const token = cookies.hanko;

  if (!token) {
    return null;
  }

  const JWKS = createRemoteJWKSet(
    new URL(`${hankoApiUrl}/.well-known/jwks.json`)
  );
  let payload;

  try {
    const verifiedJWT = await jwtVerify(token, JWKS);
    payload = verifiedJWT.payload;
  } catch {
    return null;
  }

  const userID = payload.sub;

  if (!userID) {
    return null;
  }

  return userID;
}
```

> This is how you can use it in your routes

Reviewer: ```suggestion
This is how you can use it in your routes:
```
